### PR TITLE
[octavia-ingress-controller] support wildcard in the host field

### LINKED ISSUE #2245: [Octavia-ingress-controller] Support wildcard supported by the host field in the ingress spec
/kind feature

Kubernetes ingress support wildcard at the beginning of the `host` field: https://kubernetes.io/docs/concepts/services-networking/ingress/#hostname-wildcards

The `octavia-ingress-controller` does not support wildcard today.

PS; I should have quick PR soon for that.

### PR DESCRIPTION
**Pull request by gmauleon**
**What this PR does / why we need it**:
Support wildcard for the `host` field in `ingress` resource

**Which issue this PR fixes(if applicable)**:
fixes #2245

**Special notes for reviewers**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/ingress/controller/controller.go [hunk 1]
```diff
@@ -824,10 +824,17 @@ func (c *Controller) ensureIngress(ing *nwv1.Ingress) error {
 			var policyRules []l7policies.CreateRuleOpts
 
 			if host != "" {
+
+				// Escape dot to be regex proof
+				hostRegex := strings.ReplaceAll(host, ".", "\\.")
+
+				// Handle wildcard
+				hostRegex = strings.ReplaceAll(hostRegex, "*", ".*")
```
Comment (mdbooth) [inline pkg/ingress/controller/controller.go:832]: I haven't looked at the full context, but at first glance this doesn't seem to do what it's supposed to. I'm assuming (again: sorry I haven't actually checked properly!) that the user writes `*.example.com`. They expect that to match `foo.example.com`, but do they expect it to match `foo.bar.example.com`? I suspect not, but this looks like it would match it.

File: pkg/ingress/controller/controller.go [hunk 2]
```diff
@@ -824,10 +824,17 @@ func (c *Controller) ensureIngress(ing *nwv1.Ingress) error {
 			var policyRules []l7policies.CreateRuleOpts
 
 			if host != "" {
+
+				// Escape dot to be regex proof
+				hostRegex := strings.ReplaceAll(host, ".", "\\.")
```
Comment (mdbooth) [inline pkg/ingress/controller/controller.go:829]: I realise you've just moved this here, but if the goal is to escape a string which will later be interpreted by a regexp parser (which one, btw?) then this isn't it.

**Review by mdbooth — COMMENTED** [review]
This feels like one of those areas where almost everything is a security landmine. I wonder if there's a library out there which does this safely?

At the very least there has to be something to safely escape a string literal for use in a regexp.

File: pkg/ingress/controller/controller.go (hunk 1 above)
Comment (gmauleon) [reply pkg/ingress/controller/controller.go:832]: Fair point I'll see what I can do about that.

File: pkg/ingress/controller/controller.go (hunk 2 above)
Comment (gmauleon) [reply pkg/ingress/controller/controller.go:829]: I'm not sure I follow, this was the existing "escaping" in place already and I suppose it was that simple because the `ingress` resource, sanitize its host field via API constraints, more precisely:
> a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\*\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

If I only consider wildcard at the beginning of the host` string and sanitize the rest with something like https://pkg.go.dev/regexp#QuoteMeta, would that be acceptable?

File: .gitignore [hunk 3]
```diff
@@ -159,3 +160,7 @@ zz_generated.openapi.go
 
 # local gopath
 .go
+
+### Okteto ###
+okteto.yml
+.stignore
```
Comment (jichenjc) [inline .gitignore:166]: I have no idea why we need add this?

File: pkg/ingress/controller/controller.go [hunk 4]
```diff
@@ -824,10 +825,22 @@ func (c *Controller) ensureIngress(ing *nwv1.Ingress) error {
 			var policyRules []l7policies.CreateRuleOpts
 
 			if host != "" {
+
+				hostRegex := "^"
```
Comment (jichenjc) [inline pkg/ingress/controller/controller.go:829]: I think we need a comment to describe (maybe links as well) to describe why we do like this, maybe sample as well?

File: pkg/ingress/controller/controller.go (hunk 2 above)
Comment (mdbooth) [reply pkg/ingress/controller/controller.go]: Should have been clearer in my comment: I know you didn't add this line; you just moved it, so this isn't a problem introduced by this PR.

It did make me notice it, though, which is why I commented!

File: .gitignore [hunk 5]
```diff
@@ -24,6 +24,7 @@ vendor
 
 # Vscode files
 .vscode
+__debug_bin
```
Comment (mdbooth) [inline .gitignore:27]: Same as for the Okteto lines below. Could we pull these into a separate PR?

**Comment by mdbooth** [comment]
For reference, here's the [relevant part of the definition of host](https://github.com/kubernetes/kubernetes/blob/189fe3f3e6d157bc63c2276f79c3de860d7d8c57/pkg/apis/networking/types.go#LL457C1-L468C1):
```golang
        // host can be "precise" which is a domain name without the terminating dot of
        // a network host (e.g. "foo.bar.com") or "wildcard", which is a domain name
        // prefixed with a single wildcard label (e.g. "*.foo.com").
        // The wildcard character '*' must appear by itself as the first DNS label and
        // matches only a single label. You cannot have a wildcard label by itself (e.g. Host == "*").
        // Requests will be matched against the host field in the following way:
        // 1. If host is precise, the request matches this rule if the http host header is equal to Host.
        // 2. If host is a wildcard, then the request matches this rule if the http host header
        // is to equal to the suffix (removing the first label) of the wildcard rule.
        // +optional
        Host string
```

File: pkg/ingress/controller/controller.go (hunk 4 above)
Comment (mdbooth) [reply pkg/ingress/controller/controller.go:829]: Links to:
* Octavia rule creation API: https://docs.openstack.org/api-ref/load-balancer/v2/?expanded=create-an-l7-rule-detail#create-an-l7-rule
* Ingress Rule.Host: https://github.com/kubernetes/kubernetes/blob/189fe3f3e6d157bc63c2276f79c3de860d7d8c57/pkg/apis/networking/types.go#L444-L467C2

This would also be super simple to unit test if we pulled it into a small separate function, and it's probably complicated enough to deserve it.

**Review by mdbooth — COMMENTED** [review]
Looks good to me, but agree with Ji's comments about comments and testing.
